### ISSUE TITLE
Introduce Orama internal ID for documents

### ISSUE THREAD

**Issue by micheleriva** [oramasearch/orama#426]
# Introduction to the problem

As per the Orama documentation ([link](https://docs.oramasearch.com/usage/insert#custom-document-ids)):

> Orama automatically uses the id field of the document, if found.
>
> That means that given the following document and schema:
> ```js
> import { create, search } from '@orama/orama'
> 
> const db = await create({
>   schema: {
>     id: 'string',
>     author: 'string',
>     quote: 'string',
>   },
> })
>
> await insert(db, {
>   id: '73cbcc79-2203-49b8-bb52-60d8e9a66c5f',
>   author: 'Fernando Pessoa',
>   quote: "I wasn't meant for reality, but life came and found me",
> })
>```
>
> The document will be indexed with the following `id`: `73cbcc79-2203-49b8-bb52-60d8e9a66c5f`.
>
> If the `id` field is not found, Orama will generate a random `id` for the document.

This gives users a great opportunity to use their own custom IDs, but at a cost: Orama uses several data structures (AVL Trees, Radix Trees, Inverted Indexes) where the ID of a document gets duplicated multiple times.

For example, if we insert the following document:

```js
{
  id: '37fd6bb7-5ac2-4a37-adcd-485045b54bdc',
  author: 'Michele',
  quote: 'Hello, world!'
}
```

The id `37fd6bb7-5ac2-4a37-adcd-485045b54bdc` will get duplicated at least three times:

1. `'Michele'` will be stored in the radix tree, where the last node will contain the reference of the document containing this specific token
2. `'Hello'` will be stored in the radix tree, where the last node will contain the reference of the document containing this specific token
3. `'world'` will be stored in the radix tree, where the last node will contain the reference of the document containing this specific token

There might be other places where this ID can get duplicated depending on certain conditions.

# How to solve

If the user is indexing the content using a `UUID`, this will drastically affect the index size, because of several duplications of a quite large string. 

Therefore, we should let users index every document with their own custom IDs (without putting them into a Radix Tree, therefore making the `id` property not searchable), but use an internal, shorter ID, possibly using the `syncUniqueId` function exported [here](https://github.com/oramasearch/orama/blob/main/packages/orama/src/utils.ts#L92) to store the document reference in our data structures.

Users should be able to retrieve their docs using the `getById` function ([link](https://docs.oramasearch.com/usage/utilities#getbyid)), but internally, Orama should always use a short, optimized ID.

# Bounty Program

This issue is subject to our **Open Source Bounty Program**, and we'll reward whoever is creating a PR that gets merged with $800 for this activity.

**Comment by Zizico2** [comment]
Would it make sense to keep the requirement that `id` must be a `string`? Or since there'll be an internal `id`, could that restriction be lifted?

**Comment by micheleriva** [comment]
@Zizico2 good point. I think we could allow IDs to be numbers, as long as they're unique

**Comment by Eomm** [comment]
Thinking out loud: could the id node in the tree be and object?

My idea would be to create a single class NodeId instance for every item

Than you need to keep track of the NodeId reference (instead of cloning the string id)

Finally the toString and toJson implementation should to the rest of the work

**Comment by micheleriva** [comment]
@Eomm just to clarify and see if I understood correctly:

- `id` is not a tree node but a unique string (or a number) that's part of the `Node` object (see [here](https://github.com/oramasearch/orama/blob/main/packages/orama/src/trees/radix.ts#L5))
- We could use a JS object to represent the ID if needed, but only if we can guarantee its access by reference (to avoid duplicates)
- We can't use classes as they need to be converted into different formats to be serialized and persisted to disk, which drastically increases cold starts

So, to summarize, if we want to use objects I'm ok with that, as long as we avoid duplications. It's still unclear how to access a given document by internal ID, if this one is an object

**Comment by Eomm** [comment]
> It's still unclear how to access a given document by internal ID, if this one is an object

No, this technique would solve the duplication on the other trees.
In the example the `id` is stored for each tree where the strings are added: `michele`, `hello` and `world`.
Those strings would have a reference to the `id-object` instead, meanwhile the `getById` would keep the duplicated string.

But maybe I'm just not too much into the core to understand if it would work

**Comment by H4ad** [comment]
If we do some search, should the ID returned in the document be the internal ID or the original ID?

If we need to return the original ID, I don't think any implementation will consume less memory than directly using a string.

**Comment by H4ad** [comment]
Also, I did some memory benchmarks to show how much memory could be saved if we deduplicate the IDs, you can find them here: https://github.com/H4ad/reducing-duplicate-string-memory-usage#instances-3

**Comment by micheleriva** [comment]
@H4ad the ID returned from the search should be the original ID. It shouldn't be indexed in a tree, but rather in a simple hashmap I think

**Comment by akhill10** [comment]
Just thinking about an approach: We can maintain a hashMap, so whenever a document is been created we also create an autogenerated short `_id` and map the with original ID. We will use this `_id` for all our internal purposes such as putting them into a Radix Tree, Inverted Index, etc. We can do a look-up in the hashMap to retrieve the original ID and document.

Coming to the generation of the `_id` we can make use of using the uniqueId function that already exists [here](https://github.com/oramasearch/orama/blob/main/packages/orama/src/utils.ts#L92) as mentioned in the issue  Or we can take inspiration from the elastic search on generation of those [ids](https://github.com/elastic/elasticsearch/blob/99f88f15c5febbca2d13b5b5fda27b844153bf1a/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java#L80-L118)

 /attempt #426

**Comment by Zizico2** [comment]
Loose thoughts:

- Would there need to be 2 maps? Since we would need the mapping in both directions: `internal -> original` and `original -> internal`. This exists: [bimap](https://www.npmjs.com/package/@rimbu/bimap)
- Where would the map/maps be stored? `Index` object?
- Would internal ids have to be strings? Why not numbers?
- Should the variable `lastId` used in `uniqueId()`, be stored in the `Index` object? It would make `lastId` be consistent even across serialization/de-serialization. Then we could drop the `baseId` section of the generated ids, making the ids even smaller.

What do you think @akhill10?

**Comment by H4ad** [comment]
@micheleriva If we need to return the original ID, we need to:

- have a reverse map `id -> string`, increasing memory usage (completely eliminating the potential savings).
- or iterate through the map of `string -> id` to find the correct IDs, which adds a linear search on every search.

What do you prefer?
